File: week09/25/maxprofit.cpp
// you can use includes, for example:
// #include <algorithm>

// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;

#include <vector>
#include <algorithm>
using namespace std;

int solution(vector<int> &A) {
    // Edge case: 배열이 비어 있거나 요소가 하나뿐인 경우 이익을 얻을 수 없음
    if (A.size() < 2) return 0;

    int maxProfit = 0;
    int minPrice = A[0];

    for (size_t i = 1; i < A.size(); ++i) {
        // 각 날의 이익을 계산 (현재 가격 - 최소 구매 가격)
        int profit = A[i] - minPrice;

        // 최대 이익 업데이트
        maxProfit = max(maxProfit, profit);

        // 최소 구매 가격 업데이트
        minPrice = min(minPrice, A[i]);
    }

    return maxProfit;
}

File: week09/29/maxProfit.py
# A[j] - A[i] (j>=i) 최대값 찾기
# Optimize to O(N)

def solution(A):
    if not A: return 0

    max_profit = 0
    min_price = A[0]

    for price in A[1:]:
        profit = price - min_price
        max_profit = max(max_profit, profit)
        min_price = min(min_price, price)
    
    return max_profit 

    

A = [23171, 21011, 21123, 21366, 21013, 21367]
print(solution(A))

